### ISSUE TITLE
About FSDP

### ISSUE THREAD

**Issue by pokameng** [pytorch/pytorch#108976]
@svenstaro 
I would like to ask why my GPU memory usage will be lower than FSDP mode when I am using DP mode?
model = DP(model)
`model = FSDP(model,auto_wrap_policy=my_auto_wrap_policy,
                        # use_orig_params=True,#是否保留原始参数
                        #  cpu_offload=CPUOffload(offload_params=True),
                        #  mixed_precision=fp16_policy,
                        # sharding_strategy=ShardingStrategy.SHARD_GRAD_OP, # ZeRO 2
                        sharding_strategy=ShardingStrategy.FULL_SHARD, # ZeRO 3
                        backward_prefetch = BackwardPrefetch.BACKWARD_POST, # optimize memory
                        # backward_prefetch = BackwardPrefetch.BACKWARD_PRE, # optimize speed
                        )`

**Comment by awgu** [comment]
We need more information to help.

**Comment by pokameng** [comment]
> We need more information to help.
hello, Here's my memory usage under different strategies:
![image](https://github.com/pytorch/pytorch/assets/47957015/4ee913ae-3f60-4f63-8a41-4bd7bfa933bc)
My code :

`
model = FSDP(model,auto_wrap_policy=my_auto_wrap_policy,
                        # use_orig_params=True,#是否保留原始参数
                        #  cpu_offload=CPUOffload(offload_params=True),
                        #  mixed_precision=fp16_policy,
                        # sharding_strategy=ShardingStrategy.SHARD_GRAD_OP, # ZeRO 2
                        sharding_strategy=ShardingStrategy.FULL_SHARD, # ZeRO 3
                        backward_prefetch = BackwardPrefetch.BACKWARD_POST, # optimize memory
                        # backward_prefetch = BackwardPrefetch.BACKWARD_PRE, # optimize speed
                        )
    
    
   
    loguru_logger.info("Parameter Count: %d" % count_parameters(model)) # 12659389

    if cfg.restore_ckpt is not None:
        print("[Loading ckpt from {}]".format(cfg.restore_ckpt))
        model.load_state_dict(torch.load(cfg.restore_ckpt), strict=True)

    # model.cuda()
    model.train()
    
    train_loader = datasets.fetch_dataloader(cfg)
    optimizer, scheduler = fetch_optimizer(model, cfg.trainer)

    total_steps = 0
    scaler = GradScaler(enabled=cfg.mixed_precision)
    logger = Logger(model, scheduler, cfg)
    
    reporter = MemReporter(model)
    reporter.report()
    print(f"After model loading - Memory Allocated: {torch.cuda.memory_allocated() / 1024 ** 2} MB")
    should_keep_training = True
    while should_keep_training:

        for i_batch, data_blob in enumerate(train_loader):
            
            # optimizer.zero_grad()
            images, flows, valids = [x.cuda() for x in data_blob]
            model.zero_grad(set_to_none=True)
            print(f"Before forward pass - Memory Allocated: {torch.cuda.memory_allocated() / 1024 ** 2} MB")
            
            if cfg.add_noise:
                stdv = np.random.uniform(0.0, 5.0)
                images = (images + stdv * torch.randn(*images.shape).cuda()).clamp(0.0, 255.0)

            output = {}
            flow_predictions = model(images, output)
            print(f"After forward pass (before backward) - Memory Allocated: {torch.cuda.memory_allocated() / 1024 ** 2} MB")
            loss, metrics, _ = loss_func(flow_predictions, flows, valids, cfg)
            loss.backward()
            # scaler.scale(loss).backward()
            print(f"After backward - Memory Allocated: {torch.cuda.memory_allocated() / 1024 ** 2} MB")
            scaler.unscale_(optimizer)
            print(f"After optimizer step and params update - Memory Allocated: {torch.cuda.memory_allocated() / 1024 ** 2} MB")
            torch.nn.utils.clip_grad_norm_(model.parameters(), cfg.trainer.clip)
            
            # scaler.step(optimizer)
            optimizer.step()
            scheduler.step()
            # scaler.update()

            metrics.update(output)
            logger.push(metrics)

            if total_steps % cfg.val_freq == cfg.val_freq - 1:
                PATH = '%s/%d_%s.pth' % (cfg.log_dir, total_steps+1, cfg.name)
                # torch.save(model.state_dict(), PATH)

                results = {}
                for val_dataset in cfg.validation:
                    if val_dataset == 'sintel_train':
                        results.update(evaluate_tile.validate_sintel(model.module))

                logger.write_dict(results)
                
                model.train()
            
            total_steps += 1

            if total_steps > cfg.trainer.num_steps:
                should_keep_training = False
                break

    logger.close()
    save_path = cfg.log_dir + '/final.pth'
    print(save_path)
    torch.save(model.state_dict(), save_path)

    return save_path
`

**Comment by pokameng** [comment]
> We need more information to help.

I compared the memory changes under different training strategies (DP,DDP and FSDP)

But when I set batch size =2 ,FSDP is OOM. So I'm looking for solutions like auto_wrap_bn or checkpoint_wrapper.

**Comment by pokameng** [comment]
> We need more information to help.

hello
Currently my model has about 50MB of parameters, which is not a big model. But my inputs (RGB+Optical flow) are big, can FSDP reduce the GPU memory usage during training?

**Comment by awgu** [comment]
@pokameng In that case, FSDP is not a good fit since it shards parameters, gradients, and optimizer states, which are proportional to the model size (e.g. 50 MB in your case).

You can consider activation checkpointing or microbatching in your case!

**Comment by pokameng** [comment]
Thanks,
Can you share a&nbsp;tutorial&nbsp; on activation checkpointing or microbatching with me?
I have used activation checkpointing from 【torch.distributed.algorithms._checkpoint.checkpoint_wrapper】,but GPU&nbsp; memory seems not decrease during training


繁华落尽
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "pytorch/pytorch"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年9月13日(星期三) 中午12:32
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [pytorch/pytorch] About FSDP (Issue #108976)





 
@pokameng In that case, FSDP is not a good fit since it shards parameters, gradients, and optimizer states, which are proportional to the model size (e.g. 50 MB in your case).
 
You can consider activation checkpointing or microbatching in your case!
